src/include/random.h

```diff
@@ -0,0 +1,389 @@
+}
+
+namespace detail {
+
+template <typename IntegerT, typename MutexT>
+int generate_random_number(const IntegerT max,
+                           MutexT& m,
+                           ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
+{
+  constexpr IntegerT zero = 0;   // no variable templates in C++11 :-(
+  return detail::generate_random_number<IntegerT, decltype(m),
+                                        std::uniform_int_distribution<IntegerT>,
+                                        std::default_random_engine>
+                                       (zero, max, m);
+}
+
+template <typename IntegerT, typename MutexT, typename EngineT>
+int generate_random_number(const IntegerT min, const IntegerT max,
+                           MutexT& m, EngineT& e,
+                           ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
+{
+  return detail::generate_random_number<IntegerT, MutexT,
+                                        std::uniform_int_distribution<IntegerT>,
+                                        EngineT>
+                                       (min, max, m, e);
+}
+
+template <typename IntegerT, typename MutexT, typename EngineT>
+int generate_random_number(const IntegerT max,
+                           MutexT& m, EngineT& e,
+                           ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
+{
+  constexpr IntegerT zero = 0;
+  return generate_random_number(zero, max, m, e);
+}
+
+} // namespace detail
+
+template <typename IntegerT>
+int generate_random_number(const IntegerT max,
+                           ceph::enable_if_t<std::is_integral<IntegerT>::value>* = nullptr)
+{
+  constexpr IntegerT zero = 0;   
+  return generate_random_number(zero, max);
+}
+
+namespace detail {
+
+template <typename RealT, typename MutexT>
+RealT generate_random_number(const RealT min, const RealT max, MutexT& m,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  return detail::generate_random_number<RealT, decltype(m),
+                                        std::uniform_real_distribution<RealT>, 
+                                        std::default_random_engine>
+                                       (min, max, m);
+}
+
+} // namespace detail
+
+template <typename RealT>
+RealT generate_random_number(const RealT min, const RealT max, 
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  return detail::generate_random_number<RealT,
+                                        std::uniform_real_distribution<RealT>, 
+                                        std::default_random_engine>
+                                       (min, max);
+}
+
+namespace detail {
+
+template <typename RealT, typename MutexT>
+RealT generate_random_number(const RealT max, MutexT& m,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
```

Comment:
there are still a few (unused?) overloads that mix a mutex argument with a thread-local engine